pkg/epp/datalayer/datasource.go

```diff
@@ -0,0 +1,211 @@
+limitations under the License.
+*/
+
+package datalayer
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"sync"
+)
+
+// DataSource is an interface required from all datalayer data collection
+// sources.
+type DataSource interface {
+	// Name returns the name of this datasource.
+	Name() string
+
+	// Start begins the collection process.
+	Start(ctx context.Context) error
+
+	// Stop stops the collection process.
+	Stop() error
+
+	// OutputType returns the type of information collected.
+	// TODO: this could be made private and excluded from the interface.
+	//       Use will likely be in the context of debugging/logging.
+	OutputType() reflect.Type
+
+	// AddExtractor adds an extractor to the data source.
+	// The extractor will be called whenever the data source might
+	// have some new raw information regarding an endpoint.
+	// The Extractor's expected input type is validated when it is
+	// registered.
+	AddExtractor(extractor Extractor) error
+
+	// TODO: the following is useful for a data source that operates on
+	// endpoints and might not be relevant for "global/system" collectors which
+	// might not need the concept of an endpoint. This can be split, if needed,
+	// to a separate interface in the future.
+
+	// AddEndpoint adds an endpoint to collect from.
+	AddEndpoint(ep Endpoint) error
+
+	// RemoveEndpoint removes an endpoint from collection.
+	RemoveEndpoint(ep Endpoint) error
+}
+
+type Extractor interface {
+	// Name returns the name of the extractor.
+	Name() string
+
+	// ExpectedType defines the type expected by the extractor. It must match
+	// the DataSource.OutputType() the extractor registers for.
+	ExpectedType() reflect.Type
+
+	// Extract transforms the data source output into a concrete attribute that
+	// is stored on the given endpoint.
+	Extract(data any, ep Endpoint)
+}
+
+var (
+	// DefaultDataSources is the system default data source registry.
+	DefaultDataSources = DataSourceRegistry{}
+)
+
+// DataSourceRegistry stores named data sources and makes them
+// accessible to GIE subsystems.
+type DataSourceRegistry struct {
+	mu      sync.RWMutex
+	sources map[string]DataSource
+}
+
+// Register adds a source to the registry.
+func (dsr *DataSourceRegistry) Register(src DataSource) error {
+	if src == nil {
+		return errors.New("unable to register a nil data source")
+	}
+
+	dsr.mu.Lock()
+	defer dsr.mu.Unlock()
+
+	if _, found := dsr.sources[src.Name()]; found {
+		return fmt.Errorf("unable to register duplicate data source: %s", src.Name())
+	}
+	dsr.sources[src.Name()] = src
+	return nil
+}
+
+// GetNamedSource returns the named data source, if found.
+func (dsr *DataSourceRegistry) GetNamedSource(name string) (DataSource, error) {
+	if name == "" {
+		return nil, errors.New("unable to retrieve a data source without a name")
+	}
+
+	dsr.mu.RLock()
+	defer dsr.mu.RUnlock()
+	if ds, found := dsr.sources[name]; found {
+		return ds, nil
+	}
+	return nil, &NotFoundError{Name: name}
```

Comment:
nit: consider return (DataSource, bool) instead of (DataSource, error)

Reply:
Sure. Chose an error return in case we would need something more descriptive, but as it stands now the bool is just as expressive.